File: src/main/java/com/cfgglobal/test/cache/RedisCacheClient.java
package com.cfgglobal.test.cache;

import io.vavr.control.Try;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.Set;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

import static io.vavr.API.*;
import static io.vavr.Patterns.$Failure;
import static io.vavr.Patterns.$Success;

@Component
public class RedisCacheClient implements CacheClient {
    private static Logger logger = LoggerFactory.getLogger(RedisCacheClient.class);

    @Autowired
    private RedisTemplate<String, Object> template;


    @Override
    @SuppressWarnings("unchecked")
    public <T> T get(String key) {
        return (T) template.opsForValue().get(key);
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T> T get(String key, Supplier<T> supplier) {
        Try<Object> cache = Try.of(() -> template.opsForValue().get(key));
        Optional value = Match(cache).of(
                Case($Success($()), Optional::ofNullable),
                Case($Failure($()), result -> {
                    logger.error("Query key [{}] from cache failure", key, result.getCause());
                    return Optional.empty();
                })
        );
        return (T) value.orElseGet(() -> {
            T val = null;
            try {
                val = supplier.get();
            } catch (Exception e) {
                logger.error("cannot find cached configuration {}, {}", key, e.getMessage());
            }

            logger.debug("Query key [{}] from cache success, value [{}] ", key, val);
            template.opsForValue().set(key, val);
            return val;
        });
    }

    @Override
    @SuppressWarnings("unchecked")
    public <T> T get(String key, Supplier<T> supplier, long expire) {
        Object value = get(key, supplier);
        template.expire(key, expire, TimeUnit.SECONDS);
        return (T) value;
    }


    @Override
    public void set(String key, Object value) {
        template.opsForValue().set(key, value);
    }


    @Override
    public void set(String key, long expire, Object value) {
        set(key, value);
        template.expire(key, expire, TimeUnit.SECONDS);
    }

    @Override
    public void deleteByKey(String key) {
        template.delete(key);
    }

    @Override
    public void deleteByPattern(String pattern) {
        Set<String> keys = template.keys(pattern);
        for (String key : keys) {
            logger.debug("Delete key [{}] in cache", key);
            template.delete(key);
        }
    }

    @Override
    public Set<String> keys(String pattern) {
        return template.keys(pattern);
    }


}

File: src/main/kotlin/com/github/leon/cache/RedisCacheClient.kt
package com.github.leon.cache

import arrow.core.Option
import arrow.core.getOrElse
import arrow.core.Try
import arrow.core.toOption
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit


@Component
class RedisCacheClient : CacheClient {

    @Autowired
    private val template: RedisTemplate<String, Any>? = null


    override fun <T> get(key: String): T? {
        return template!!.opsForValue().get(key) as T?
    }

    override fun <T> get(key: String, supplier: () -> T): T? {
        val cache = Try { template!!.opsForValue().get(key) as T? }
        return when (cache) {
            is Try.Success -> cache.value.toOption()
            is Try.Failure -> {
                logger.error("Query key [{}] from cache failure", key, cache.exception.message)
                Option.empty()
            }
        }.getOrElse {
            val value =  supplier()
            logger.debug("Query key [{}] from cache success, value [{}] ", key, value)
            template!!.opsForValue().set(key, value)
            value
        }
    }

    override fun <T> get(key: String, supplier: () -> T, expire: Long): T? {
        val value = get(key, supplier)
        template!!.expire(key, expire, TimeUnit.SECONDS)
        return value
    }


    override fun set(key: String, value: Any) {
        template!!.opsForValue().set(key, value)
    }


    override fun set(key: String, expire: Long, value: Any) {
        set(key, value)
        template!!.expire(key, expire, TimeUnit.SECONDS)
    }

    override fun deleteByKey(key: String) {
        template!!.delete(key)
    }

    override fun deleteByPattern(pattern: String) {
        val keys = template!!.keys(pattern)
        for (key in keys) {
            logger.debug("Delete key [{}] in cache", key)
            template.delete(key)
        }
    }

    override fun keys(pattern: String): Set<String> {
        return template!!.keys(pattern)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(com.github.leon.cache.RedisCacheClient::class.java)
    }


}
